modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -96,38 +102,37 @@ public IncomingSnapshotCopier(PartitionSnapshotStorage partitionSnapshotStorage,
     public void start() {
         Executor executor = partitionSnapshotStorage.getIncomingSnapshotsExecutor();
 
-        LOG.info("Copier is started for the partition [partId={}, tableId={}]", partId(), tableId());
+        LOG.info("Copier is started for the partition [{}]", createPartitionInfo());
+
+        joinFuture = new CompletableFuture<>();
 
-        future = prepareMvPartitionStorageForRebalance()
-                .thenCompose(unused -> prepareTxStatePartitionStorageForRebalance(executor))
+        rebalanceFuture = partitionSnapshotStorage.partition().startRebalance()
                 .thenCompose(unused -> {
                     ClusterNode snapshotSender = getSnapshotSender(snapshotUri.nodeName);
 
                     if (snapshotSender == null) {
-                        LOG.error(
-                                "Snapshot sender not found [partId={}, tableId={}, nodeName={}]",
-                                partId(),
-                                tableId(),
-                                snapshotUri.nodeName
-                        );
-
-                        if (!isOk()) {
-                            setError(RaftError.UNKNOWN, "Sender node was not found or it is offline");
-                        }
-
-                        return completedFuture(null);
+                        throw new StorageRebalanceException("Snapshot sender not found: " + snapshotUri.nodeName);
                     }
 
                     return loadSnapshotMeta(snapshotSender)
                             .thenCompose(unused1 -> loadSnapshotMvData(snapshotSender, executor))
-                            .thenCompose(unused1 -> loadSnapshotTxData(snapshotSender, executor))
-                            .thenAcceptAsync(unused1 -> updateLastAppliedIndexFromSnapshotMetaForStorages(), executor);
+                            .thenCompose(unused1 -> loadSnapshotTxData(snapshotSender, executor));
+                });
+
+        rebalanceFuture.handle((unused, throwable) -> completesRebalance(throwable))
+                .thenCompose(Function.identity())
+                .whenComplete((unused, throwable) -> {
+                    if (throwable == null) {
+                        joinFuture.complete(null);
```

Comment:
Why do you need a second object? This is not explained and not obvious at all

Reply:
Fix it.